### PR TITLE
Fix overspecialize of literal

### PR DESCRIPTION
**Pull request by sklam**
Fix overspecialization of literals due to excessive literal dispatch. The key change is flipping the preference to use non-literal type over literal type. All other changes are for fixes to make the tests pass. 

I don't think there are any realistic test that we can add to check for the reduction in specialization. However, the https://github.com/numba/numba/pull/6035#issuecomment-665680185 provides how I have measured the changes in literal dispatch.

### CODE REVIEW

**Comment by sklam** [comment]
The PR is now only failing on:

* `numba.tests.test_lists.TestLiteralLists.test_contains`
* `numba.tests.test_dictobject.TestLiteralStrKeyDict.test_dict_as_arg`
* `numba.tests.test_dictobject.TestLiteralStrKeyDict.test_build_map_op_code`

So mostly on `LiteralList` and `LiteralStrKeyDict`.

**Comment by sklam** [comment]
Using a patch to measure the number of literal dispatch:

```diff
diff --git a/numba/core/types/functions.py b/numba/core/types/functions.py
index b4c00d389..a648b722e 100644
--- a/numba/core/types/functions.py
+++ b/numba/core/types/functions.py
@@ -299,6 +299,13 @@ class BaseFunction(Callable):
                     if sig is not None:
                         self._impl_keys[sig.args] = temp.get_impl_key(sig)
                         self._depth -= 1
+
+                        def islit(x):
+                            return isinstance(x, (types.Literal, types.InitialValue))
+                        if uselit and any(map(islit, args)):
+                            import os
+                            with open(f"lit_{os.getpid()}.log", "a") as fout:
+                                print(temp, args, kws, file=fout)
                         return sig
                     else:
                         registered_sigs = getattr(temp, 'cases', None)
```

and with the tests `python runtests.py numba/tests/test_typedlist.py numba/tests/test_dictobject.py`, the number of literal dispatch reduces from 4332 (commit d3b7ea1344779cf0bacfa82c754da5b71e133dbe) to 67 (commit 4c86201e957f29d625f083db58a384f150a27101). This shows that this PR has significantly reduce the number of literal dispatch.

**Comment by sklam** [comment]
BFID: `numba_smoketest_cpu_56`

**Comment by esc** [comment]
I checked this PR using the patch provided:

```
 💣 zsh» less lit_42112.log |  wc -l                                                                                                                              :(
67
 💣 zsh» cat lit_43621.log| wc -l
4384
```

**Comment by esc** [comment]
interestingly enough, this is now failing on Python 3.8 on `aarch64` with: the same error as: https://github.com/numba/numba/pull/5980

It's always along the lines of:

```
[2020-07-30 05:58:04,333] {docker_operator.py:262} INFO - ======================================================================
[2020-07-30 05:58:04,333] {docker_operator.py:262} INFO - ERROR: test_basic96 (numba.tests.test_stencils.TestManyStencils)
[2020-07-30 05:58:04,333] {docker_operator.py:262} INFO - 1D slice.
[2020-07-30 05:58:04,334] {docker_operator.py:262} INFO - ----------------------------------------------------------------------
[2020-07-30 05:58:04,334] {docker_operator.py:262} INFO - Traceback (most recent call last):
[2020-07-30 05:58:04,334] {docker_operator.py:262} INFO - File "/opt/conda/envs/testenv_c69c1b12-b62c-4c51-83a4-6d2590cde95f/lib/python3.8/site-packages/numba/tests/test_stencils.py", line 2759, in test_basic96
[2020-07-30 05:58:04,335] {docker_operator.py:262} INFO - self.check(kernel, a, options={'neighborhood': ((-1, 1),)})
[2020-07-30 05:58:04,335] {docker_operator.py:262} INFO - File "/opt/conda/envs/testenv_c69c1b12-b62c-4c51-83a4-6d2590cde95f/lib/python3.8/site-packages/numba/tests/test_stencils.py", line 1668, in check
[2020-07-30 05:58:04,335] {docker_operator.py:262} INFO - raise RuntimeError(str1 + str2)
[2020-07-30 05:58:04,336] {docker_operator.py:262} INFO - RuntimeError: The following implementations should not have raised an exception but did:
[2020-07-30 05:58:04,336] {docker_operator.py:262} INFO - ['pyStencil']
[2020-07-30 05:58:04,336] {docker_operator.py:262} INFO - Errors were:
[2020-07-30 05:58:04,337] {docker_operator.py:262} INFO - 
[2020-07-30 05:58:04,337] {docker_operator.py:262} INFO - pyStencil: Message: <class 'ValueError'>: Name node can't be used with 'None' constant
```

I'm starting to think something might be wrong with that machine.

**Comment by sklam** [comment]
RE: https://github.com/numba/numba/pull/6035#issuecomment-666340203

I wonder if the PyStencil error message has to do with a bug in `numba_do_raise` that i fixed in the other PR.

**Comment by esc** [comment]
@sklam but in that case it should be present also in the current build of `master`.

**Comment by esc** [comment]
I just noticed a significant difference between the smoketest and the fulltest: for `aarch64` - the smoketest does Python 3.7 and 3.8 -- but the fulltest does 3.6 and 3.7

smoketest:
Screen Shot 2020-07-30 at 14 55 13

fulltest:
Screen Shot 2020-07-30 at 14 55 48

**Comment by esc** [comment]
So I reckon it might be a 3.8 issue of some sort.

**Comment by esc** [comment]
I managed to reproduce the error with `aarch64` using the qemu/docker setup on OSX to emulate `aarch64`, I'll open an issue with a reproducer later on today.

**Comment by esc** [comment]
@sklam any thoughts on how to break this?

**Comment by esc** [comment]
With #6057 we have a likely fix for the CI failures, hence I am marking this as BF passed.

**Comment by sklam** [comment]
Tests from https://github.com/numba/numba/pull/6058/files#diff-f905c47974bdaea9bed2212720d50660R1811-R1814 allow us to test this patch by showing that the lack of `prefer_literal=True` will use non-literal types first and, thus, returning an implementation that is not specialized.

**Comment by stuartarchibald** [comment]
Runtimes for 4ea838b:
```
numba.numba Successful in 24m — Build #20200729.15 succeeded
numba.numba (Linux py36_np115_32bit) Successful in 6m — Linux py36_np115_32bit succeeded
numba.numba (Linux py36_np115_cov) Successful in 7m — Linux py36_np115_cov succeeded
numba.numba (Linux py36_np117_doc) Successful in 6m — Linux py36_np117_doc succeeded
numba.numba (Linux py37_np115) Successful in 5m — Linux py37_np115 succeeded
numba.numba (Linux py37_np115_32bit) Successful in 7m — Linux py37_np115_32bit succeeded
numba.numba (Linux py37_np116_omp) Successful in 5m — Linux py37_np116_omp succeeded
numba.numba (Linux py37_np116_svml) Successful in 4m — Linux py37_np116_svml succeeded
numba.numba (Linux py37_np116_tbb) Successful in 7m — Linux py37_np116_tbb succeeded
numba.numba (Linux py37_np116_vanilla) Successful in 5m — Linux py37_np116_vanilla succeeded
numba.numba (Linux py37_np116_workqueue) Successful in 6m — Linux py37_np116_workqueue succeeded
numba.numba (Linux py37_np117_pickle5) Successful in 5m — Linux py37_np117_pickle5 succeeded
numba.numba (Linux py37_np118) Successful in 5m — Linux py37_np118 succeeded
numba.numba (Linux py38_np117) Successful in 4m — Linux py38_np117 succeeded
numba.numba (Linux py38_np118) Successful in 4m — Linux py38_np118 succeeded
numba.numba (Windows py37_np115) Successful in 11m — Windows py37_np115 succeeded
numba.numba (Windows py38_np118) Successful in 10m — Windows py38_np118 succeeded
numba.numba (macOS py37_np116) Successful in 23m — macOS py37_np116 succeeded
numba.numba (macOS py38_np117) Successful in 7m — macOS py38_np117 succeeded
```
will rerun, the variation is large, looks like:
```
numba.numba (macOS py37_np116) Successful in 23m — macOS py37_np116 succeeded
```
got all the list tests, it may be necessary to split these into literal and non-literal if it reproduces.

**Comment by stuartarchibald** [comment]
Second run looks like:
```
numba.numba Successful in 5m — Build #20200803.6 succeeded
numba.numba (Linux py36_np115_32bit) Successful in 7m — Linux py36_np115_32bit succeeded
numba.numba (Linux py36_np115_cov) Successful in 7m — Linux py36_np115_cov succeeded
numba.numba (Linux py36_np117_doc) Successful in 6m — Linux py36_np117_doc succeeded
numba.numba (Linux py37_np115) Successful in 5m — Linux py37_np115 succeeded
numba.numba (Linux py37_np115_32bit) Successful in 5m — Linux py37_np115_32bit succeeded
numba.numba (Linux py37_np116_omp) Successful in 5m — Linux py37_np116_omp succeeded
numba.numba (Linux py37_np116_svml) Successful in 5m — Linux py37_np116_svml succeeded
numba.numba (Linux py37_np116_tbb) Successful in 6m — Linux py37_np116_tbb succeeded
numba.numba (Linux py37_np116_vanilla) Successful in 4m — Linux py37_np116_vanilla succeeded
numba.numba (Linux py37_np116_workqueue) Successful in 5m — Linux py37_np116_workqueue succeeded
numba.numba (Linux py37_np117_pickle5) Successful in 7m — Linux py37_np117_pickle5 succeeded
numba.numba (Linux py37_np118) Successful in 5m — Linux py37_np118 succeeded
numba.numba (Linux py38_np117) Successful in 6m — Linux py38_np117 succeeded
numba.numba (Linux py38_np118) Successful in 6m — Linux py38_np118 succeeded
numba.numba (Windows py37_np115) Successful in 15m — Windows py37_np115 succeeded
numba.numba (Windows py38_np118) Successful in 14m — Windows py38_np118 succeeded
numba.numba (macOS py37_np116) Successful in 5m — macOS py37_np116 succeeded
numba.numba (macOS py38_np117) Successful in 5m — macOS py38_np117 succeeded
```
guess the issue was transient.

File: numba/core/types/functions.py [hunk 1]
```diff
@@ -224,6 +224,16 @@ def raise_error(self):
         raise errors.TypingError(self.format())
 
 
+def _unlit_non_poison(ty):
+    """Apply unliteral(ty) and raise TypingError if type is a Poison.
+    """
+    out = unliteral(ty)
+    if isinstance(out, types.Poison):
+        m = "Poison type used in arguments"
+        raise TypingError(m)
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/core/types/functions.py]: added in https://github.com/numba/numba/pull/6035/commits/8acb58db54134ffba7d2bf199e887c9f02850324

**Comment by esc** [comment]
@sklam thank you for addressing all the reviewer comments, I am marking this as RTM.

**Pull request merged**